src/Host/Client/Impl/Host/RHost.Message.cs

```diff
@@ -12,32 +16,68 @@ private class Message {
 
+            public ConcurrentQueue<byte[]> Blobs;
+            public readonly int ExpectedBlobs;
+
             private readonly JArray _body;
             private readonly int _argsOffset;
 
             public int ArgumentCount => _body.Count - _argsOffset;
 
             public Message(JToken token) {
                 _body = token as JArray;
+                var header = _body[0] as JArray;
+
                 if (_body == null) {
                     throw ProtocolError($"Message must be an array:", token);
                 }
-                if (_body.Count < 2) {
-                    throw ProtocolError($"Message must have form [id, name, ...]:", token);
+
+                if(header == null) {
+                    throw ProtocolError($"Message header must be an array:", header);
+                }
+
+                if (_body.Count < 1) {
+                    throw ProtocolError($"Message must have form [[id, name, blob_count], ...]:", token);
+                }
+
+                if (header.Count < 3) {
+                    throw ProtocolError($"Message header must have form [[id, name, blob_count], ...]:", header);
+                }
+
+                var id = header[0];
+                if(id.Type != JTokenType.String) {
+                    throw ProtocolError($"id must be {JTokenType.String}:", this);
+                }
+                Id = (string)id;
+
+                var name = header[1];
+                if (name.Type != JTokenType.String) {
+                    throw ProtocolError($"name must be {JTokenType.String}:", this);
                 }
+                Name = (string)name;
 
-                Id = GetString(0, "id");
-                Name = GetString(1, "name");
-                _argsOffset = 2;
+                var blob = header[2];
+                if (blob.Type != JTokenType.Integer) {
+                    throw ProtocolError($"blob_count must be {JTokenType.Integer}:", this);
+                }
+                ExpectedBlobs = (int)blob;
 
                 if (Name.StartsWithOrdinal(":")) {
-                    if (_body.Count < 3) {
-                        throw ProtocolError($"Response message must have form [id, name, request_id, ...]:", token);
+                    if (header.Count < 4) {
+                        throw ProtocolError($"Response message must have form [[id, name, blob_count, request_id, ...], ...]:", token);
                     }
 
-                    RequestId = GetString(0, "request_id");
-                    ++_argsOffset;
+                    var requestId = header[3];
+                    if (requestId.Type != JTokenType.String) {
+                        throw ProtocolError($"request_id must be {JTokenType.String}:", this);
+                    }
+                    RequestId = (string)requestId;
                 }
+
+                // header part is done
+                _argsOffset = 1;
+
+                Blobs = new ConcurrentQueue<byte[]>(); 
```

Comment:
We need to parse the header before we know how many blob (slices) we need to receive. I have a task on my plate to refactor some of this code. I will add a note to expose blobs as `IReadOnlyCollection<byte[]>`